[New Concept]: Case (#486)
exercism/crystal: concepts/case/about.md

Case allows for an optional else statement which is executed if no other case matches.

The case expression is evaluated and then compared to each when expression.
The expression is compared using the [case subsumption operator (`===`)][case-subsumption].

```Crystal
case 1
when 1
  puts "One"
when 2
  puts "Two"
else
  puts "Other"
end
# => One

# This is the same as:
tmp = 1
if 1 === tmp
  puts "One"
elsif 1 === tmp
  puts "Two"
else
  puts "Other"
end
```

## Case subsumption operator (`===`)

The case subsumption operator (`===`) is a bit different from the equality operator (`==`).
The operator checks if the right side is a member of the set described by the left side.
This means that it does matter where each operand is placed.
How this works depends on the type of the left side, for example a `Range` would check if the right side is in the range or a `Object` would check if the right side is an instance of the `Object`.

```Crystal
(1..3) == 1  # => false
(1..3) === 1 # => true

String == "foo"  # => false
String === "foo" # => true
```

## Case with multiple expressions

> Cases allows for matching multiple expressions in a single case, each possible value is separated by a comma.

Reviewer: ```suggestion
Cases allow for matching multiple expressions in a single case with each possible value separated by a comma.
```

It will execute the code if any of the expressions match.
This can be useful when you want a single case to have multiple possible values.

```Crystal
case 1
when 1, 2
  puts "One or two"
else
  puts "Other"
end
```

## Cases with ranges

Cases can also check if a value is in a range.
This is done by having a range as the when expression.

```Crystal
case var
when 1..3
  puts "One to three"
else
  puts "Other"
end
```

## Cases with no case expression

> When there is no need for a case expression then it is possible to omit it.

Reviewer: ```suggestion
When there is no need for a case expression, it is possible to omit it.
```

> Doing this will make so each case expression is evaluated for truthiness.

Reviewer: ```suggestion
Doing this will make it so that each case expression is evaluated for truthiness.
```

```Crystal
case var
when 1 then puts "One"
when 2 then puts "Two"
else puts "Other"
end
```

## Case with types

Case allows for the matching with types.
The comparison will use the `is_a?` operator to check the type of the variable.
This allows for example to have one branch for each type in a union type.

```Crystal
case var
when Int32
  puts "Int32"
when String
  puts "String"
else
  puts "Other"
end
```

## Implicit object

Case allows for the use of an implicit object.
This allows for example a when expression to be a method call on the case expression.
Based on if the method returns a truthy or falsey value will the case be matched or not.

```Crystal
case 1
when .odd?
  puts "Odd"
when .even?
  puts "Even"
end
```

## Exhaustive case

Case allows for exhaustive cases, this means that it is possible to ensure that all possible values are handled.
It is written similar to the when statement, but with a `in` keyword instead of `when`.
When a case is exhaustive, so does it not support any `when` or `else` statements.

Exhaustive cases can be used on `Bools`, `Union` types, `Enum`, and `Tuples`.
The two latter ones will be covered in later concepts.

> Exhaustive cases means that all possibilities will have to be handled, otherwise will a compile error be raised.

Reviewer: ```suggestion
Exhaustive cases means that all possibilities will have to be handled, otherwise a compile error will be raised.
```
